oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -169,25 +189,113 @@ impl Runtime {
         // for any additional work to be finished here. This may take an arbitrary amount of time,
         // depending on the work that the node thread has to perform, but at least we know that the
         // it will not be able to enter again in a blocking state.
-        for (_, mut node) in nodes.drain() {
-            node.instance.stop();
+        // for mut instance in instances.drain(..) {
+        for mut instance in instances {
+            instance.stop();
+        }
+    }
+
+    /// Allow the corresponding [`Node`] to use the [`Handle`]s passed via the iterator.
+    /// This is achieved by adding the [`Handle`]s to the [`Node`]s [`HashMap`] of [`Handle`]s.
+    fn track_handles_in_node<I>(&self, node_id: NodeId, handles: I)
+    where
+        I: IntoIterator<Item = Handle>,
+    {
+        if node_id == RUNTIME_NODE_ID {
+            return;
+        }
+
+        let nodes = self.nodes.read().unwrap();
+        let node = nodes
+            .get(&node_id)
+            .expect("Invalid node_id passed into track_handles_in_node!");
+
+        let mut tracked_handles = node.handles.lock().unwrap();
+        for handle in handles {
+            tracked_handles.insert(handle);
+        }
+    }
+
+    /// Replace [`Handle`] with [`None`] if the [`NodeId`] does not have access to the [`Handle`].
+    fn validate_handle_access(&self, node_id: NodeId, handle: Handle) -> Result<(), OakStatus> {
+        // Allow RUNTIME_NODE_ID access to all handles.
+        if node_id == RUNTIME_NODE_ID {
+            return Ok(());
+        }
+
+        let nodes = self.nodes.read().unwrap();
+        // Lookup the node_id in the runtime's nodes hashmap
+        let node = nodes
+            .get(&node_id)
+            .expect("Invalid node_id passed into validate_handle_access!");
+        let tracked_handles = node.handles.lock().unwrap();
+
+        // Check the handle exists in the handles associated with a node, otherwise
+        // return None.
+        if tracked_handles.contains(&handle) {
+            Ok(())
+        } else {
+            error!(
+                "validate_handle_access: handle {:?} not found in node {:?}",
+                handle, node_id
+            );
+            Err(OakStatus::ErrBadHandle)
+        }
+    }
+
+    /// Replace [`Handle`]s with [`None`] if the [`NodeId`] does not have access to the [`Handle`]s.
+    /// Any iterator is accepted, but a Vec<Option<Handle>> is returned so as not to hold
+    /// underlying Mutexes open.
+    fn validate_handles_access<'a, I>(&self, node_id: NodeId, handles: I) -> Result<(), OakStatus>
+    where
+        I: Iterator<Item = &'a Option<Handle>>,
```

Comment:
Consider `IntoIterator` here too.

I also don't quite remember what the point of the `Option` is inside the iterator, could you clarify in the comment?

Reply:
IntoIterator and Option removed in #787